pkg/gadgettracermanager/containerutils/containerutils.go

```diff
@@ -172,6 +172,25 @@ func PidFromContainerId(containerID string) (int, error) {
 			return -1, fmt.Errorf("invalid pid")
 		}
 		return crioInspect.Pid, nil
+	} else if strings.HasPrefix(containerID, "containerd://") {
+		out, err := exec.Command("chroot", "/host", "crictl", "inspect", strings.TrimPrefix(containerID, "containerd://")).Output()
+		if err != nil {
+			return -1, err
+		}
+		type CRIOInspect struct {
```

Comment:
This seems to be the same code as the previous "if" block, it could be factored together (except the TrimPrefix, which could be replaced with 
https ://golang.org/pkg/regexp/#Regexp.Split).

The name of the type definition and variable should be changed since this is not about cri-o but about crictl.

Reply:
The code is very similar but the structure type used in both cases is different:

cri-o:
```
type CRIOInspect struct {
	Pid int
}
```
containerd: (I'll update to the correct name

```
type CRIOInspect struct {
  Info struct {
    Pid int
  }
}
```

I think this format is dependent on the runtime and not on the crictl version but I haven't found a confirmation. Do you know more about it?

Reply:
ctrctl's `outputStatusInfo()` function marshalls the `info` as an embedded struct manually:
https://github.com/kubernetes-sigs/cri-tools/blob/c6533b12a128079eedd35b364e17033c34feab58/cmd/crictl/util.go#L220-L231

I don't think it is a matter of cri-o versus containerd. It's probably a different version of ctrctl on Ubuntu that does the marshalling differently. What version do you have?

I think we should call the gRPC method ourselves instead of parsing the output of the command line. That would work because the gRPC proto seems more stable than the marshalling of the crictl tool. Also, it would work when crictl is not installed on the host.

For this PR, maybe just add a `// TODO: call the gRPC method ContainerStatus() directly instead of parsing the output of the crictl tool`.

-----

crictl sends the gRPC `ContainerStatus()` method and it returns a `ContainerStatusResponse`: [crictl/container.go#L749](https://github.com/kubernetes-sigs/cri-tools/blob/7c03c059b403cced3b12c62c5c1c11c5b0797193/cmd/crictl/container.go#L749)

ContainerStatusResponse contains 2 fields ([api.proto#L944-L952](https://github.com/kubernetes/cri-api/blob/82de05bf5a58d97ede91ad8faa68d61d0e9d0f07/pkg/apis/runtime/v1alpha2/api.proto#L944-L952)):
- `ContainerStatus status`
- `map<string, string> info`, which contains the pid.

Both containerd and cri-o uses this `info` map which contains the pid. However, for some reason, the serialisation in json string is different.

Here (on the OpenShift cluster) we can see that the `info` field is present but is marshalled as an embedded struct:
```
# crictl inspect eeb8e4eb15259 -o table
ID: eeb8e4eb15259d79f7a886298952256bdb2265d4e67b3e79e134c7214edd3511
Name: gadget
State: CONTAINER_RUNNING
Created: 4 minutes ago
Started: 4 minutes ago
Labels:
        io.kubernetes.container.name -> gadget
        io.kubernetes.pod.name -> gadget-f8ks8
        io.kubernetes.pod.namespace -> kube-system
        io.kubernetes.pod.uid -> 21799f5c-a653-11ea-b966-0242ac11000a
Annotations:
        io.kubernetes.container.hash -> e1559af5
        io.kubernetes.container.preStopHandler -> {"exec":{"command":["/cleanup.sh"]}}
        io.kubernetes.container.restartCount -> 0
        io.kubernetes.container.terminationMessagePath -> /dev/termination-log
        io.kubernetes.container.terminationMessagePolicy -> File
        io.kubernetes.pod.terminationGracePeriod -> 30
Info: map[pid:146265 sandboxId:f29f6625a677a0835123fa6ac201d198fb8d0cd92704a6d7a6596f93213975f4]
```

```
# crictl inspect eeb8e4eb15259 -o json|tail -3
  "pid": 146265,
  "sandboxId": "f29f6625a677a0835123fa6ac201d198fb8d0cd92704a6d7a6596f93213975f4"
}
```